core/capabilities/remote/dispatcher.go

```diff
@@ -0,0 +1,146 @@
+package remote
+
+import (
+	"context"
+	"fmt"
+	sync "sync"
+
+	"google.golang.org/protobuf/proto"
+
+	"github.com/smartcontractkit/chainlink-common/pkg/services"
+	commontypes "github.com/smartcontractkit/chainlink-common/pkg/types"
+
+	remotetypes "github.com/smartcontractkit/chainlink/v2/core/capabilities/remote/types"
+	"github.com/smartcontractkit/chainlink/v2/core/logger"
+	p2ptypes "github.com/smartcontractkit/chainlink/v2/core/services/p2p/types"
+)
+
+// dispatcher en/decodes messages and routes traffic between peers and capabilities
+type dispatcher struct {
+	peerWrapper p2ptypes.PeerWrapper
+	peer        p2ptypes.Peer
+	peerID      p2ptypes.PeerID
+	signer      p2ptypes.Signer
+	registry    commontypes.CapabilitiesRegistry
+	receivers   map[key]remotetypes.Receiver
+	mu          sync.RWMutex
+	stopCh      services.StopChan
+	wg          sync.WaitGroup
+	lggr        logger.Logger
+}
+
+type key struct {
+	capId string
+	donId string
+}
+
+var _ services.Service = &dispatcher{}
+
+const supportedVersion = 1
+
+func NewDispatcher(peerWrapper p2ptypes.PeerWrapper, signer p2ptypes.Signer, registry commontypes.CapabilitiesRegistry, lggr logger.Logger) *dispatcher {
+	return &dispatcher{
+		peerWrapper: peerWrapper,
+		signer:      signer,
+		registry:    registry,
+		receivers:   make(map[key]remotetypes.Receiver),
+		stopCh:      make(services.StopChan),
+		lggr:        lggr.Named("Dispatcher"),
+	}
+}
+
+func (d *dispatcher) Start(ctx context.Context) error {
+	d.peer = d.peerWrapper.GetPeer()
```

Comment:
This feels like some tight coupling -- can we just pass in d.peerWrapper.GetPeer() at instantiation time?

Reply:
As discussed offline - Peer depends on unlocked Keystore so it can't be fully initialized on init. Technically I could merge Peer and PeerWrapper into a single object but I prefer to keep Peer agnostic of any core-specific components such as Keystore.

This setup is very similar to SingletonPeerWrapper.